### ISSUE TITLE
Version data repositories not using Git but using sound patch system

### ISSUE THREAD

**Issue by strogonoff** [paneron/paneron#1]
Considering adopting a version control system based on patches (Pijul/Darcs type).

It offers a huge advantage in that you can e.g. unapply a change made in the past without altering the history and having to force-push, since the identity of a patch once created never changes.

This would give us a reliable backbone for managing revisions of objects (1) without making us conform to rigid history of Git and its unreliable merging algorithms, and (2) without having to duplicate object revision management in actual file contents (which breaks layers of abstraction).

To migrate, we can write converters for different repository types between our “YAML-with-revisions” convention and Pijul repository format.

**Comment by ronaldtse** [comment]
I agree that the Darcs/Pijul property of reversing a patch anywhere to be very attractive. However the popularity of the Git platform, which is used by all our users, remain important where they can deploy and launch their registries...

**Comment by strogonoff** [comment]
Your comment is about GitHub not Git

**Comment by ronaldtse** [comment]
The comment was about Git. While we consider integrating with GitHub, the other platforms like GitLab and BitBucket are also relevant.

**Comment by strogonoff** [comment]
Sure.

Would you agree if there was a PijulHub, it would not be an issue?

Thus the concern appears to be about managing collaboration-facilitating infrastructure for organizations in our care (i.e. we can’t rely on GitHub for free), not solution as it stands on its own.

Judging on the merits for the solution itself, so far I believe YAML with revisions on top of Git is redundant, fragile and less DRY, losing to YAML on top of a versioning system based on superior patch theory.

It needs to be taken into account. We should ask, what would we do if there was no GitHub? And apply the same logic here.

**Comment by ronaldtse** [comment]
@strogonoff I'm not sure if a PijulHub obsoletes the issue.

* A register has a need to track register items
* Each register item has an individual history
* We want to store a register in a data repository (hopefully those repositories are readily useable by users)

I agree that the Pijul/Darcs model works closest to this. However, the data administrators are familiar with files (in an archive) or Git because they can be considered a portable data archive. 

I also agree that implementing revisions on top of Git is redundant and fragile. It would be better if the model of the data operations align with the operations of the register.
